engine/src/main/scala/cromwell/engine/workflow/CallCachingBlacklistManager.scala

```diff
@@ -0,0 +1,132 @@
+package cromwell.engine.workflow
+
+import akka.event.LoggingAdapter
+import com.google.common.cache.{CacheBuilder, CacheLoader, LoadingCache}
+import com.typesafe.config.Config
+import cromwell.backend.standard.callcaching.{BlacklistCache, GroupingBlacklistCache, RootWorkflowBlacklistCache}
+import cromwell.core.CacheConfig
+import mouse.boolean._
+import net.ceedubs.ficus.Ficus._
+
+import scala.concurrent.duration._
+import scala.language.postfixOps
+
+object CallCachingBlacklistManager {
+  object Defaults {
+    object Groupings {
+      val Concurrency = 1000
+      val Size = 1000L
+      val Ttl = 2 hours
+    }
+    object Hits {
+      val Concurrency = 1000
+      val Size = 20000L
+      val Ttl = 1 hour
+    }
+    object Buckets {
+      val Concurrency = 1000
+      val Size = 1000L
+      val Ttl = 1 hour
+    }
+  }
+}
+
+class CallCachingBlacklistManager(rootConfig: Config, logger: LoggingAdapter) {
+
+  // Defined if "call-caching.blacklist-cache.enabled = true".
+  private val blacklistCacheConfig: Option[Unit] =
+    rootConfig.getOrElse("call-caching.blacklist-cache.enabled", false).option(())
+
+  // Defined if `blacklistCacheConfig` is defined and "call-caching.blacklist-cache.groupings.workflow-option" is defined.
+  private val blacklistGroupingWorkflowOptionKey: Option[String] = for {
+    _ <- blacklistCacheConfig // Only return a groupings cache if blacklisting is enabled.
+    workflowOption <- rootConfig.as[Option[String]]("call-caching.blacklist-cache.groupings.workflow-option")
+  } yield workflowOption
+
+  // Defined if `blacklistGroupingWorkflowOptionKey` is defined.
+  private val blacklistGroupingCacheConfig: Option[CacheConfig] = {
+    import CallCachingBlacklistManager.Defaults.Groupings._
+    for {
+      _ <- blacklistGroupingWorkflowOptionKey
+      groupingsOption = rootConfig.as[Option[Config]] ("call-caching.blacklist-cache.groupings")
+      conf = CacheConfig.config(groupingsOption, defaultConcurrency = Concurrency, defaultSize = Size, defaultTtl = Ttl)
+    } yield conf
+  }
+
+  // Defined if `blacklistCacheConfig` is defined.
+  private val blacklistBucketCacheConfig: Option[CacheConfig] = {
+    import CallCachingBlacklistManager.Defaults.Buckets._
+    for {
+      _ <- blacklistCacheConfig
+      bucketsOption = rootConfig.as[Option[Config]]("call-caching.blacklist-cache.buckets")
+      conf = CacheConfig.config(bucketsOption, defaultConcurrency = Concurrency, defaultSize = Size, defaultTtl = Ttl)
+    } yield conf
+  }
+
+  // Defined if `blacklistCacheConfig` is defined.
+  private val blacklistHitCacheConfig: Option[CacheConfig] = {
+    import CallCachingBlacklistManager.Defaults.Hits._
+    for {
+      _ <- blacklistCacheConfig
```

Comment:
This long long chain of `for`-comprehensions... could it all be done in a single one?

Reply:
no, some of these things can be defined independently of one another.